### DISCUSSION TITLE
How to Handle `net6.0`-only Code

### DISCUSSION THREAD

**Discussion by TheCodeTraveler** [CommunityToolkit/Maui#301] [Q&A] [answered]
## Question

What is the best way to add support for `net6.0`-only code?

## Current Setup

In `CommunityToolkit.Maui.csproj` and `CommunityToolkit.Maui.Core.csproj`, we specify specific behavior for the following file types:
- `*.shared.cs`
- `*.ios.cs`
- `*.macos.cs`
- `*.macios.cs`
- `*.android.cs`
- `*.windows.cs`

The platform-specific files only compile + run on that specific platform (with the exception of `macios` which runs on both iOS + MacCatalyst), and the `*.shared.cs` files compile + run on each platform.

## Problem

The current setup leaves us without an option for code that _only_ runs on `net6.0` (ie not a platform like `net6.0-ios` or `net6.0-android`), aka `net6.0`-only code.

Currently, we handle `net6.0`-only code by using the following ifdef in a `*.shared.cs` file:

```cs
#if !(IOS || ANDROID || MACCATALYST || WINDOWS)
  // net6.0 code
#endif
```

Here is an example from `Snackbar.shared.cs`:

https://github.com/CommunityToolkit/Maui/blob/2fba2f4fb8297a507f065ab2aae58e1fe52aa675/src/CommunityToolkit.Maui/Alerts/Snackbar/SnackBar.shared.cs#L103-L123

**Comment by TheCodeTraveler** [comment]
### Proposal

I propose that, in `CommunityToolkit.Maui.csproj` and `CommunityToolkit.Maui.Core.csproj`, we add one more file type:
- `*.net.cs`

```xml
  <ItemGroup Condition="$(TargetFramework.StartsWith('net')) != true AND '$(TargetFramework)' != 'net6.0'">
    <Compile Remove="**\*.net" />
    <None Include="**\*.net" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" />
    <Compile Remove="**\net\**\*.cs" />
    <None Include="**\net\**\*.cs" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" />
  </ItemGroup>
```

This would then allow us to create partial classes for `net6.0`-only code, e.g. `Snackbar.net.cs`.

## Unknowns / Concerns

Before implementing this, I have one major concern:

How will this affect .NET MAUI apps that don't use a single project + multi-targeting?

If a .NET MAUI app has a `net6.0` project where they write all of their cross-platform .NET MAUI code (similar to how we use .NET Standard for Xamarin.Forms apps), will the `net6.0` project defer to the `*.android.cs` file when compiled + run? Or will it compile + run the `*.net.cs` code instead? 

If the `*.net.cs` code is chosen, it will throw errors such as `NotImplementedException` when the `PopupViewHandler` runs `CreateNativeElement()`

This likely won't be a common scenario, but it is a valid .NET architecture that we should take into account to avoid future problems.

**Reply by bijington** [reply]
Could we compile all `*.shared.cs` files for `net6.0` and not need to include this additional partial class?

Disclaimer: I haven't considered the implications of doing this yet and whether it would a) solve our problem and b) even be possible

**Reply by pictos** [reply]
@bijington doing that will cause some compile issues, I tried that, making all .net6 on `shared` and that makes the compiler crazy. But I would like to remove the `shared` suffix for our files :D

**Reply by bijington** [reply]
Is that expected behaviour by the compiler?

**Reply by pictos** [reply]
I think so

**Reply by VladislavAntonyuk** [reply]
maybe we can do it with Constants?
```
<PropertyGroup Condition="$(TargetFramework)==(IOS || ANDROID || MACCATALYST || WINDOWS)">
  <DefineConstants>Platofrms</DefineConstants>
</PropertyGroup>
```

**Reply by maxkoshevoi** [reply]
+1 for just manually defining some new symbol. For now consensus is probably on `NET_ANY`
Here's another possible solution for this: https://github.com/dotnet/sdk/issues/22788#issuecomment-1010220030

> will the `net6.0` project defer to the `*.android.cs` file when compiled + run?

Interesting question. From my understanding of MSBuild both approaches (`#if` or `*.net.cs`) should work the same, since both of them simply exclude/include some code from being compiled.

**Reply by pictos** [reply]
Don't like the idea of having just one way to handle that. `#if` are really good when you have just a small portion of code that will run on only .net6 or in a specific scenario. But for most scenarios `#if` will make the code hard to read and maintain, so having a file for that will be the best. We can take as reference projects like Xamarin.Essentials and .NET MAUI, they use a file to handle netstandard (aka .net6) files and they succeed.

**Reply by VladislavAntonyuk** [reply]
Agree with @pictos.
For the Snackbar example I would prefer #if. As for the big changes like in Popup I would go with *.net.cs

**Reply by maxkoshevoi** [reply]
> We can take as reference projects like Xamarin.Essentials and .NET MAUI, they use a file to handle netstandard (aka .net6) files and they succeed.

If they are doing it, then Toolkit should follow to keep style consistent. 

I was coming from an idea that `net` code is usually pretty minimal (couple of lines), and maybe there's no need to a separate file in most cases (and short `#if` will do). But if there are more then two-three lines of code, separate file makes sense.

**Comment by VladislavAntonyuk** [comment, marked as answer]
I believe this one is resolved by https://github.com/CommunityToolkit/Maui/pull/576

**Comment by CliffAgius** [comment]
Closed as answered
